Firestore/Source/Remote/FSTSerializerBeta.mm

```diff
@@ -1144,8 +1148,13 @@ - (FSTDocumentWatchChange *)decodedDocumentChange:(GCFSDocumentChange *)change {
   const DocumentKey key = [self decodedDocumentKey:change.document.name];
   SnapshotVersion version = [self decodedVersion:change.document.updateTime];
   HARD_ASSERT(version != SnapshotVersion::None(), "Got a document change with no snapshot version");
-  FSTMaybeDocument *document =
-      [FSTDocument documentWithData:value key:key version:version state:FSTDocumentStateSynced];
+  // The document may soon be re-serialized back to protos in order to store it in local
+  // persistence. Memoize the encoded form to avoid encoding it again.
```

Comment:
FWIW, we made this choice intentionally, knowing about that doc. In short:

  * Clients are tightly coupled to the server version anyway because there's so much in the way of semantics tied up in the representation and much of our storage is really just queueing requests for the backend
  * We're depending on a stable public API rather than an unstable internal one
  * So long as the field tags and types line up we can rename these into private types later

In the steady state we can trust there won't be breaking changes to the protos because unlike google3 we can't impose on customers that they upgrade.

The only point where there is some diceyness is around transitions between versions of the protocol. The transition from v1alpha1 to v1beta1 was grossly incompatible, but thankfully at the time we didn't have persistence yet. Note that v1alpha1 was essentially the Cloud Datastore Entity model with streaming bolted on, while v1beta1 was Firestore proper so this was an expected breaking change. The upcoming transition to v1 is a no-op because v1beta1 and v1 are wire compatible.

Even if the RPC protocol were to change radically we can likely insist that the Document part remain stable across versions. Regardless, a fork-when-we-need-to strategy has allowed us to defer this work for two years and counting so I'm going to keep betting on it :-).